Source/WebCore/dom/SerializedNode.cpp

```diff
@@ -29,16 +29,74 @@
 #include "Attr.h"
 #include "CDATASection.h"
 #include "Comment.h"
+#include "Document.h"
 #include "DocumentType.h"
+#include "HTMLAttachmentElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLInputElement.h"
+#include "HTMLScriptElement.h"
+#include "HTMLSourceElement.h"
 #include "JSNode.h"
 #include "ProcessingInstruction.h"
+#include "QualifiedName.h"
+#include "SVGScriptElement.h"
 #include "SecurityOriginPolicy.h"
+#include "SliderThumbElement.h"
 #include "Text.h"
+#include "WebVTTElement.h"
 
 namespace WebCore {
 
-JSC::JSValue SerializedNode::deserialize(SerializedNode&& serializedNode, JSC::JSGlobalObject* lexicalGlobalObject, JSDOMGlobalObject* domGlobalObject, WebCore::Document& document)
+WTF_MAKE_STRUCT_TZONE_ALLOCATED_IMPL(SerializedElement);
+
+static void deserializeAssociatedAttachmentElement(std::unique_ptr<SerializedElement>&& attachmentAssociatedElement, AttachmentAssociatedElement& element, Document& document)
+{
+#if ENABLE(ATTACHMENT_ELEMENT)
+    if (RefPtr attachment = attachmentAssociatedElement ? dynamicDowncast<HTMLAttachmentElement>(SerializedElement::deserialize(WTFMove(*attachmentAssociatedElement), document)) : nullptr)
+        element.setAttachmentElement(*attachment);
+#endif
+}
+
+SerializedElement::AttachmentAssociatedElement::AttachmentAssociatedElement(std::unique_ptr<SerializedElement>&& attachmentAssociatedElement)
+    : attachmentAssociatedElement(WTFMove(attachmentAssociatedElement)) { }
+
+SerializedElement::AttachmentAssociatedElement::AttachmentAssociatedElement(const AttachmentAssociatedElement& other)
+    : attachmentAssociatedElement(other.attachmentAssociatedElement ? makeUnique<SerializedElement>(*other.attachmentAssociatedElement) : nullptr) { }
+
+Ref<Element> SerializedElement::deserialize(SerializedElement&& element, WebCore::Document& document)
 {
+    constexpr bool createdByParser { false };
+    return WTF::switchOn(WTFMove(element.data), [&] (SerializedElement::Element&& element) -> Ref<WebCore::Element> {
+        return document.createElement(WTFMove(element.name).qualifiedName(), createdByParser);
+    }, [&] (SerializedElement::SliderThumbElement&&) -> Ref<WebCore::Element> {
+        return WebCore::SliderThumbElement::create(document);
+    }, [&] (SerializedElement::WebVTTElement&& element) -> Ref<WebCore::Element> {
+        return WebCore::WebVTTElement::create(element.type, AtomString(WTFMove(element.language)), document);
+    }, [&] (SerializedElement::HTMLImageElement&& element) -> Ref<WebCore::Element> {
+        Ref deserialized = WebCore::HTMLImageElement::create(document);
+        deserializeAssociatedAttachmentElement(WTFMove(element.attachmentAssociatedElement), deserialized, document);
```

Comment:
I'm pretty sure we don't need to support attachment element.
This feature is only used in Mail's compose window.